src/Storages/MergeTree/MutateTask.cpp

```diff
@@ -2321,6 +2420,12 @@ bool MutateTask::prepare()
         /// Keeper has to be asked with unlock request to release the references to the blobs
         ctx->new_data_part->remove_tmp_policy = IMergeTreeDataPart::BlobsRemovalPolicyForTemporaryParts::ASK_KEEPER;
 
+        bool drop_expired_parts = suitable_for_ttl_optimization && !ctx->data->getSettings()->materialize_ttl_recalculate_only;
+        if (drop_expired_parts)
+            task = std::make_unique<ExecutableTaskDropTTLExpiredPartsDecorator>(std::make_unique<MutateSomePartColumnsTask>(ctx), ctx);
+        else
+            task = std::make_unique<MutateSomePartColumnsTask>(ctx);
+
         task = std::make_unique<MutateSomePartColumnsTask>(ctx);
```

Comment:
And here.

Reply:
Sorry, I really didn't get it. What is wrong with this and the one above?

Reply:
The line `task = std::make_unique<MutateSomePartColumnsTask>(ctx);` is always called and overrides the task. Therefore your previous code with `if-else` is not effective.